api/src/main/java/ai/djl/nn/transformer/ScaledDotProductAttentionBlock.java

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.nn.transformer;
+
+import ai.djl.ndarray.NDArray;
+import ai.djl.ndarray.NDList;
+import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+import ai.djl.nn.Block;
+import ai.djl.nn.core.Linear;
+import ai.djl.nn.norm.Dropout;
+import ai.djl.training.ParameterStore;
+import ai.djl.util.PairList;
+
+/**
+ * A Block implementing scaled product attention according to <a
+ * href="https://arxiv.org/abs/1706.03762">Vaswani et. al.</a>.
+ *
+ * <p>Abbreviations used:
+ *
+ * <ul>
+ *   <li>E = embedding size
+ *   <li>B = batch size
+ *   <li>N = number of attention heads
+ *   <li>F = "from" sequence length (key/value sequence), the input sequence
+ *   <li>T = "to" sequence length (query sequence), e.g. the length of the output sequence
+ *   <li>S = a sequence length, either F or T
+ *   <li>H = Attention head size (= E / N)
+ * </ul>
+ *
+ * <p>In many use cases F=T. For self attention, the input is equal to the output.
+ *
+ * <p>This block can process input in four forms:
+ *
+ * <ul>
+ *   <li>Input size one: [Values] = [(B, F, E)], only input is used as key, query and value
+ *       (unmasked self attention), e.g. BERT
+ *   <li>Input size two: [Values, Mask] = [(B, F, E), (B, F, F)], first input is used as key, query
+ *       and value, masked self attention
+ *   <li>Input size three: [Keys, Queries, Values] = [(B, F, E), (B, T, E), (B, F, E)], inputs are
+ *       interpreted as keys, queries and values, unmasked attention
+ *   <li>Input size four: [Keys, Queries, Values, Mask] = [(B, F, E), (B, T, E), (B, F, E), (B, T,
+ *       F)], inputs are interpreted as keys, queries, values and an attention mask, full masked
+ *       attention.
+ * </ul>
+ *
+ * <p>Attention masks must contain a 1 for positions to keep and a 0 for positions to mask.
+ */
+public final class ScaledDotProductAttentionBlock extends TransformerBaseBlock {
+
+    private static final int VERSION = 1;
+
+    /** Size of the Word-/Token-embeddings we use the attention on. */
+    private final int embeddingSize;
```

Comment:
Just FYI: We don't encourage add final for member/local and parameter variable. It's OK the leave the final, but it's a bit verbose and doesn't provide much value.

Reply:
I fixed continuous build issue, rebase latest code should pass the tests.

Reply:
I did a merge this time as I already merged before and didn't want to run into conflicts due to the rebase. I will work with rebases instead of merges for future PRs.

Reply:
As far as the `final` modifier is concerened: You are right, in this case the benefits are not so huge. On whether it generally provides much value - I am afraid we have to agree to disagree here ;) 
I consider using final on members, parameters and local variables extensively one of the best coding practices. It helps avoid a lot of problems that come with mutable state and IMHO it encourages an immutable, more bug-resilient style. However, I will limit my use of final to more tricky control flow and classes with a more state and a more complex lifecycle. I removed the modifier in this case so the new Block fits in better with the existing code. It is a pity that the `final` behaviour isn't the default in Java - I wish there was  a `mutable` keyword instead...